modules/miner/mine.go

```diff
@@ -45,6 +41,64 @@ func (m *Miner) blockForWork() (b types.Block) {
 	return
 }
 
+// solveBlock takes a block, target, and number of iterations as input and
+// tries to find a block that meets the target. This function can take a long
+// time to complete, and should not be called with a lock.
+func (m *Miner) solveBlock(blockForWork types.Block, target types.Target, iterations uint64) (b types.Block, solved bool, err error) {
+	b = blockForWork
+	bRoot := b.MerkleRoot()
+	hashbytes := make([]byte, 72)
+	copy(hashbytes, b.ParentID[:])
+	copy(hashbytes[40:], bRoot[:])
+
+	// Iterate through a bunch of nonces (from a random starting point) and try
+	// to find a winnning solution.
+	nonce := (*uint64)(unsafe.Pointer(&hashbytes[32]))
+	*nonce = b.Nonce
+	for i := 0; i < int(iterations); i++ {
+		*nonce++
+		id := crypto.HashBytes(hashbytes)
+		if bytes.Compare(target[:], id[:]) >= 0 {
+			b.Nonce = binary.LittleEndian.Uint64(hashbytes[32:])
+			err = m.cs.AcceptBlock(b)
+			if err != nil {
+				println("Mined a bad block " + err.Error())
+				m.tpool.PurgeTransactionPool()
+			}
+			solved = true
+			if build.Release != "testing" {
+				println("Found a block! Reward will be received in 50 blocks.")
+			}
+
+			// Grab a new address for the miner.
+			m.mu.Lock()
+			var addr types.UnlockHash
+			addr, _, err = m.wallet.CoinAddress()
+			if err == nil { // Special case: only update the address if there was no error.
+				m.address = addr
+			}
+			m.mu.Unlock()
+			return
+		}
+	}
+
+	return
+}
+
+// increaseAttempts is the miner's way of guaging it's own hashrate. After it's
+// made 100 attempts to find a block, it calculates a hashrate based on how
+// much time has passed. The number of attempts in progress is set to 0
+// whenever mining starts or stops, which prevents weird low values from
+// cropping up.
+func (m *Miner) increaseAttempts() {
+	m.attempts++
+	if m.attempts >= 100 {
+		m.hashRate = int64((m.attempts * m.iterationsPerAttempt * 1e9)) / (time.Now().UnixNano() - m.startTime)
+		m.startTime = time.Now().UnixNano()
+		m.attempts = 0
```

Comment:
I think you can avoid calling `.UnixNano()` all over the place. Just store it as a `time.Time` and convert where necessary.